Adding more Webots Simulation Tutorials

ygoumaz opened the pull request:
Over the past few weeks, we've received a lot of questions about connecting Webots and ROS 2, indicating a significant interest in interfacing both frameworks.

Although there is already a tutorial available on `docs.ros.org`, we have additional ones about the `webots_ros2` package, such as installation, explanations, and advanced simulation setup, which are currently only accessible through the [wiki of the corresponding Github repository](https://github.com/cyberbotics/webots_ros2/wiki). However, people seem to have difficulty locating these tutorials.

We believe that if these tutorials were included in the official ROS 2 documentation, they would be more visible. This PR proposes a new structure for simulation tutorials and adds three new pages on how to install the package on each platform. If this is accepted, we plan to add more tutorial pages in the future.

ygoumaz committed:
Update intro to WSL

ygoumaz committed:
Update intro to WSL

ygoumaz committed:
Revert "Update intro to WSL"

This reverts commit 52a52d09928ce8f1e159a667bf6e8a0e125ccf05.

ygoumaz committed:
Update intro to wsl

ygoumaz committed:
Update intro to wsl

ygoumaz committed:
Remove windows tabs

ygoumaz committed:
Remove windows tabs

ygoumaz committed:
Add windows WSL tab for extern controller code

ygoumaz committed:
Add windows WSL tab for extern controller code

ygoumaz committed:
Clean tutorial

ygoumaz committed:
Clean tutorial

ygoumaz committed:
Update world with declarations

ygoumaz committed:
Merge pull request #3 from cyberbotics/fix-update-world-to-r2022b

Update world of the tutorial to R2022b

ygoumaz committed:
Add supervisor node and environment var

ygoumaz committed:
Add ros2supervisorlauncher node creation in launch files

ygoumaz committed:
Add ros2supervisorlauncher node creation in launch files

ygoumaz committed:
add env variable for linux

ygoumaz committed:
Update tutorial with explanations

ygoumaz committed:
Update tutorial with explanations

ygoumaz committed:
Merge branch 'rolling' into enhancement-replace-windows-compatibility-with-wsl

ygoumaz committed:
Merge branch 'rolling' into fix-start-ros2supervisor-node

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: ad-daniel <44834743+ad-daniel@users.noreply.github.com>

ygoumaz committed:
Merge pull request #4 from cyberbotics/fix-start-ros2supervisor-node

Fix missing supervisor node creation and explanation

ygoumaz committed:
Merge branch 'rolling' into enhancement-replace-windows-compatibility-with-wsl

ygoumaz committed:
include correct file for supervisor part

ygoumaz committed:
include correct file for supervisor part

ygoumaz committed:
Add WEBOTS_CONTROLLER_URL wsl infos

ygoumaz committed:
Add WEBOTS_CONTROLLER_URL wsl infos

ygoumaz committed:
cleanup

ygoumaz committed:
cleanup

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Remove confusing sentence

ygoumaz committed:
Remove confusing sentence

ygoumaz committed:
Merge branch 'enhancement-replace-windows-compatibility-with-wsl' of https://github.com/cyberbotics/ros2_documentation into enhancement-replace-windows-compatibility-with-wsl

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Merge pull request #2 from cyberbotics/enhancement-replace-windows-compatibility-with-wsl

Update Windows compatibility to WSL in tutorials

ygoumaz committed:
Squashed commit of the following:

commit 72effa715be53dcf2954bb694a430b0d8e956c8a
Author: Luke Schmitt <85308904+LSinterbotix@users.noreply.github.com>
Date:   Tue Nov 1 17:26:25 2022 -0500

    Fix list formatting on Contact page (#3153)

commit dcd066fc11b1d1248b25d0ab4003ac4dcbebcacc
Author: Dan Zimmerman <dan@danzimmerman.com>
Date:   Tue Nov 1 16:29:34 2022 -0400

    Update throttle logging docs with units and data type hints (re: rclcpp/issues/1929) (#3143)

    * Note milliseconds on API Concepts page

    * Add units and types to About Logging page

    Co-authored-by: dan <dan@polarworks.no>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>

commit 375b6cf2610bc4c3b775df8bde3f94deaa515d29
Author: definitelyuncertain <saikirann94@gmail.com>
Date:   Fri Oct 28 14:30:22 2022 -0500

    Change to mention explicit requirement for folder names for .msg and .srv files (#3144)

    * Change to mention explicit requirement for folder names for msg and srv files

    From https://github.com/ros2/tutorials/issues/29

    * fix typo

    Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

    * fix rst formatting

    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

    Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

commit da871ef458644134bdabaa6589b62767b1e97dda
Author: Sergei Sergienko <divelix1@gmail.com>
Date:   Thu Oct 27 17:20:06 2022 +0300

    fix link to Gazebo-ROS2 bridge + CLI examples (#3139)

    * fix link to Gazebo-ROS2 bridge + CLI examples

    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

commit e19c9b50fcd4851502f1e7aeca6d8c72b952b129
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Wed Oct 26 16:31:21 2022 -0400

    Update the language to match the source code. (#3135)

    Fallout from #2865

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit a5910a16cc466519ede87779a7a56796096149a3
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Wed Oct 26 08:38:36 2022 -0400

    Shorten the Windows build-from-source path. (#3130)

    * Shorten the Windows build-from-source path.

    This is to keep the total length down below the 260 character
    limit imposed by Windows by default.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 82256860b9cc6ec968893f0fe3ab3a4dcb01fcf7
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Mon Oct 24 08:30:37 2022 -0400

    Fix up the variable name in the traveling in time (C++) tutorial. (#3115)

    This is just to match what is in the code in the upstream
    repository.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 87ab121ee41a84cebec8929dfa4b541f9139ce11
Author: Ian Colwell <48106342+icolwell-as@users.noreply.github.com>
Date:   Fri Oct 21 11:07:30 2022 -0400

    Update default WORKING_DIRECTORY for ament_add_gtest() (#3117)

commit 87bf20a6097c6aa53877d36ec0e7c56ab92dfe1e
Author: chianfern <chianfern@gmail.com>
Date:   Tue Oct 18 21:36:57 2022 +0800

    Add ParameterDescriptor to 'Using parameters in a class (C++)' (#2865)

    * Rename node name in "Using parameters" C++ tutorial to be consistent with Python tutorial.

    Signed-off-by: Tan Chian Fern <chianfern@gmail.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 77646cea81c3d014d5b8a33d2709a0bb9fc64529
Author: George Stavrinos <gstavrinos@protonmail.com>
Date:   Tue Oct 18 15:29:37 2022 +0300

    Fixed Intel's robot SDK repo with the new devkit name + reference to MoveIt2 instead of 1. (#3107)

commit 1d24955992c420d6f85580bc0c7b7805769e138d
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Mon Oct 17 14:06:52 2022 -0400

    Monitor parameter changes: Add note to save callback handle (#3102)

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 8910a80ffb785264071e813c3b049e9561bf2344
Author: David V. Lu!! <davidvlu@gmail.com>
Date:   Fri Oct 14 09:56:14 2022 -0400

    Some testing basics (#3084)

    * Some testing basics

    Signed-off-by: David V. Lu <davidvlu@gmail.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 877145260f7d0282f695442971971ecc39799be6
Author: Tomoya Fujita <Tomoya.Fujita@sony.com>
Date:   Fri Oct 14 05:34:20 2022 -0700

    Sony's task for Iron Release. (#3089)

    * Sony's task for Iron Release.

    Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>

    * Rearrange and add in links.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

    Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>

commit 2ea9e7d78456edf3bf07385b65fd6c248cdc15cf
Author: Yannick Goumaz <61198661+ygoumaz@users.noreply.github.com>
Date:   Fri Oct 14 14:23:13 2022 +0200

    Update Setting-up a Robot Simulation (Webots) tutorial (#3090)

    * Update intro to WSL

    * Update world with declarations

    * Add supervisor node and environment var

    * Update tutorial with explanations

    Co-authored-by: ad-daniel <44834743+ad-daniel@users.noreply.github.com>

commit 0390a50713588f0a9477f523e30f05efe2e8cbac
Author: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
Date:   Thu Oct 13 13:42:30 2022 -0400

    fix unresolved package requirement to install Humble on Ubuntu 20.04 (#3086) (#3088)

    * fix command to install humble on Ubuntu 20.04

    Co-authored-by: paolo.dimasi <paodimasi@outlook.com>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>
    (cherry picked from commit 5626af9b050f89c9e6895f6e461cf3ec01492238)

    Co-authored-by: Paolo Dimasi <46300254+Pamasi@users.noreply.github.com>

ygoumaz committed:
Squashed commit of the following:

commit 72effa715be53dcf2954bb694a430b0d8e956c8a
Author: Luke Schmitt <85308904+LSinterbotix@users.noreply.github.com>
Date:   Tue Nov 1 17:26:25 2022 -0500

    Fix list formatting on Contact page (#3153)

commit dcd066fc11b1d1248b25d0ab4003ac4dcbebcacc
Author: Dan Zimmerman <dan@danzimmerman.com>
Date:   Tue Nov 1 16:29:34 2022 -0400

    Update throttle logging docs with units and data type hints (re: rclcpp/issues/1929) (#3143)

    * Note milliseconds on API Concepts page

    * Add units and types to About Logging page

    Co-authored-by: dan <dan@polarworks.no>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>

commit 375b6cf2610bc4c3b775df8bde3f94deaa515d29
Author: definitelyuncertain <saikirann94@gmail.com>
Date:   Fri Oct 28 14:30:22 2022 -0500

    Change to mention explicit requirement for folder names for .msg and .srv files (#3144)

    * Change to mention explicit requirement for folder names for msg and srv files

    From https://github.com/ros2/tutorials/issues/29

    * fix typo

    Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

    * fix rst formatting

    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

    Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

commit da871ef458644134bdabaa6589b62767b1e97dda
Author: Sergei Sergienko <divelix1@gmail.com>
Date:   Thu Oct 27 17:20:06 2022 +0300

    fix link to Gazebo-ROS2 bridge + CLI examples (#3139)

    * fix link to Gazebo-ROS2 bridge + CLI examples

    Co-authored-by: Chris Lalancette <clalancette@gmail.com>

commit e19c9b50fcd4851502f1e7aeca6d8c72b952b129
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Wed Oct 26 16:31:21 2022 -0400

    Update the language to match the source code. (#3135)

    Fallout from #2865

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit a5910a16cc466519ede87779a7a56796096149a3
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Wed Oct 26 08:38:36 2022 -0400

    Shorten the Windows build-from-source path. (#3130)

    * Shorten the Windows build-from-source path.

    This is to keep the total length down below the 260 character
    limit imposed by Windows by default.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 82256860b9cc6ec968893f0fe3ab3a4dcb01fcf7
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Mon Oct 24 08:30:37 2022 -0400

    Fix up the variable name in the traveling in time (C++) tutorial. (#3115)

    This is just to match what is in the code in the upstream
    repository.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 87ab121ee41a84cebec8929dfa4b541f9139ce11
Author: Ian Colwell <48106342+icolwell-as@users.noreply.github.com>
Date:   Fri Oct 21 11:07:30 2022 -0400

    Update default WORKING_DIRECTORY for ament_add_gtest() (#3117)

commit 87bf20a6097c6aa53877d36ec0e7c56ab92dfe1e
Author: chianfern <chianfern@gmail.com>
Date:   Tue Oct 18 21:36:57 2022 +0800

    Add ParameterDescriptor to 'Using parameters in a class (C++)' (#2865)

    * Rename node name in "Using parameters" C++ tutorial to be consistent with Python tutorial.

    Signed-off-by: Tan Chian Fern <chianfern@gmail.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 77646cea81c3d014d5b8a33d2709a0bb9fc64529
Author: George Stavrinos <gstavrinos@protonmail.com>
Date:   Tue Oct 18 15:29:37 2022 +0300

    Fixed Intel's robot SDK repo with the new devkit name + reference to MoveIt2 instead of 1. (#3107)

commit 1d24955992c420d6f85580bc0c7b7805769e138d
Author: Chris Lalancette <clalancette@openrobotics.org>
Date:   Mon Oct 17 14:06:52 2022 -0400

    Monitor parameter changes: Add note to save callback handle (#3102)

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 8910a80ffb785264071e813c3b049e9561bf2344
Author: David V. Lu!! <davidvlu@gmail.com>
Date:   Fri Oct 14 09:56:14 2022 -0400

    Some testing basics (#3084)

    * Some testing basics

    Signed-off-by: David V. Lu <davidvlu@gmail.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

commit 877145260f7d0282f695442971971ecc39799be6
Author: Tomoya Fujita <Tomoya.Fujita@sony.com>
Date:   Fri Oct 14 05:34:20 2022 -0700

    Sony's task for Iron Release. (#3089)

    * Sony's task for Iron Release.

    Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>

    * Rearrange and add in links.

    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

    Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>
    Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>

commit 2ea9e7d78456edf3bf07385b65fd6c248cdc15cf
Author: Yannick Goumaz <61198661+ygoumaz@users.noreply.github.com>
Date:   Fri Oct 14 14:23:13 2022 +0200

    Update Setting-up a Robot Simulation (Webots) tutorial (#3090)

    * Update intro to WSL

    * Update world with declarations

    * Add supervisor node and environment var

    * Update tutorial with explanations

    Co-authored-by: ad-daniel <44834743+ad-daniel@users.noreply.github.com>

commit 0390a50713588f0a9477f523e30f05efe2e8cbac
Author: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
Date:   Thu Oct 13 13:42:30 2022 -0400

    fix unresolved package requirement to install Humble on Ubuntu 20.04 (#3086) (#3088)

    * fix command to install humble on Ubuntu 20.04

    Co-authored-by: paolo.dimasi <paodimasi@outlook.com>
    Co-authored-by: Chris Lalancette <clalancette@openrobotics.org>
    (cherry picked from commit 5626af9b050f89c9e6895f6e461cf3ec01492238)

    Co-authored-by: Paolo Dimasi <46300254+Pamasi@users.noreply.github.com>

ygoumaz committed:
Update Webots.rst

ygoumaz committed:
Update Webots.rst

ygoumaz committed:
Update Webots.rst

ygoumaz committed:
Merge branch 'rolling' of https://github.com/cyberbotics/ros2_documentation into rolling

ygoumaz committed:
Remove trailing white spaces

ygoumaz committed:
Merge branch 'ros2:rolling' into rolling

ygoumaz committed:
Merge branch 'rolling' of https://github.com/cyberbotics/ros2_documentation into rolling

ygoumaz committed:
Update to VM and no supervisor

ygoumaz committed:
Remove wrong text

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Merge pull request #7 from cyberbotics/enhancement-rviz-and-update

Update Setting-up-a-robot-simulation to 2023.0.2

ygoumaz committed:
Merge branch 'ros2:rolling' into rolling

omichel committed:
Merge branch 'ros2:rolling' into rolling

ygoumaz committed:
Enhancement reorganize tutorials and add installation procedure (#8)

* Re-arrange files + linux & windows installation

* macOS instructions

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-MacOS.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-MacOS.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-MacOS.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-MacOS.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-MacOS.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Windows.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulators/Installation-Windows.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Reorganize folders

* Update source/Tutorials/Advanced/Simulation-Webots/Installation-Linux.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update macOS tutorial + link to image

* Update Installation-MacOS.rst

---------

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz committed:
Merge remote-tracking branch 'upstream/rolling' into rolling (#9)

ygoumaz committed:
Merge remote-tracking branch 'upstream/rolling' into rolling

ygoumaz committed:
fix trailing whitespaces

kscottz commented on source/Tutorials/Advanced/Simulation-Webots/Simulation-Webots.rst:
```diff
@@ -0,0 +1,20 @@
+Unlike turtlesim, they provide fairly realistic results relying on physics-based models for robots, sensors, actuators and objects.
+Hence, what you observe in simulation is very close to what you will get when transferring your ROS 2 controllers to a real robot.
+
+This set of tutorials will teach you how to configure the Webots simulator with ROS 2.
```
I think it would be productive here to give a brief overview of Webots and why one might want to use it over a different simulator. I wouldn't go so far as to write marketing copy, or use superlatives, but an even handed description of the application domains where Webots is commonly used. Tell the reader, who may know very little about simulators where and when Webots would be a good choice.

omichel replied:
[Here is](https://github.com/ros2/ros2_documentation/blob/94dc49a89f8657b134237a1b3e0fe7da38a8c369/source/Tutorials/Advanced/Simulation-Webots/Simulation-Webots.rst?plain=1#L8-L14) what I wrote to answer your concern. I am not very comfortable comparing Webots to Gazebo, as I am not an expert of Gazebo. However, I believe those are some significant differences a user should consider when choosing a simulator.

kscottz commented on source/Tutorials/Advanced/Simulation-Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -20,9 +21,6 @@ Setting up a robot simulation (Webots)
 Background
 ----------
 
-Several robot simulators can be used with ROS 2, such as Gazebo, Webots, etc.
```
I suggest keeping this section but moving it below line 24. It provides some context to less versed readers.

omichel replied:
This was actually moved into the [introductory page](https://github.com/ros2/ros2_documentation/blob/94dc49a89f8657b134237a1b3e0fe7da38a8c369/source/Tutorials/Advanced/Simulation-Webots/Simulation-Webots.rst?plain=1#L4) where I believe it is better located.

kscottz reviewed (commented):
I did a cursory review of this pull request. I did read through the tutorials and they look complete and well written. I can't comment on the structural changes or the functionality of the tutorials, I'll leave that up to the maintainers. My primary suggestion is that you provide more context up front about what a simulator is and why would pick one over the other (using even-handed language). 

Am I correct in assuming that this is preparation for the robot wrestling competition at ICRA?

omichel commented on source/Tutorials/Advanced/Simulation-Webots/Simulation-Webots.rst:
```diff
@@ -0,0 +1,20 @@
+Simulation [Webots]
+===================
+
+Several robot simulators can be used with ROS 2, such as Gazebo, Webots, etc.
```
```suggestion
Several advanced robot simulators can be used with ROS 2, such as Gazebo, Webots, etc.
```

omichel commented:
> Am I correct in assuming that this is preparation for the robot wrestling competition at ICRA?

Not, really, we simply want to make the Webots ROS 2 documentation better in general. For sure, it will benefit to the participants of the ICRA competition, but not only them. We have no plans to add a documentation page specific to this ICRA competition as it seems out-of-the-scope of the ROS 2 documentation pages. We already host such a documentation including example controllers on the competition repositories [here](https://github.com/cyberbotics/wrestling-alice-ros-2) and [there](https://github.com/cyberbotics/wrestling-bob-ros-2).

omichel committed:
Update Simulation-Webots.rst

ygoumaz commented:
@clalancette and @audrow, kindly pinging you, as you were removed from the reviewers.

clalancette commented on source/Tutorials/Advanced.rst:
```diff
@@ -10,5 +10,6 @@ Advanced
    Advanced/Recording-A-Bag-From-Your-Own-Node-CPP
    Advanced/Recording-A-Bag-From-Your-Own-Node-Py
-   Advanced/Simulators/Simulation-Main
+   Advanced/Simulation-Gazebo/Simulation-Gazebo
+   Advanced/Simulation-Webots/Simulation-Webots
```
I don't agree with this change.  That is, I think we should still keep the Simulation-Main page, and then have Webots and Gazebo linked off of there.  After all, there may be more Simulators added in the future, and that provides a nice structure for them to fit into.  It also means we don't need to deal with the awkward question of where to redirect `Simulation-Main` to.

clalancette commented on source/Tutorials/Advanced/Simulation-Webots/Installation-Linux.rst:
```diff
@@ -0,0 +1,131 @@
+----------
+
+The ``webots_ros2`` package provides an interface between ROS 2 and Webots.
+It includes several sub-packages, including ``webots_ros2_driver``, which allows to start Webots and communicate with it.
```
```suggestion
It includes several sub-packages, including ``webots_ros2_driver``, which allows you to start Webots and communicate with it.
```

clalancette commented on source/Tutorials/Advanced/Simulation-Webots/Installation-Linux.rst:
```diff
@@ -0,0 +1,131 @@
+Installation (Linux)
```
For almost all other tutorials, we have a single page with multiple tabs to cover the differences between Linux, macOS, and Windows.  Could a similar thing work here?  That is, instead of having three separate pages, we'd have one page with 3 tabs. 

(please consider this one a suggestion; if that makes things awkward, I'm happy to leave this as separate pages)

clalancette replied:
Separately, we should be more specific than "Linux".  Since we now support RHEL as well as Ubuntu (and Debian and Fedora unofficially), we should be specific on which Linux distribution(s) this targets.

ygoumaz replied:
We have already discussed this point on our end. Initially, we had intended to include it on our wiki page, but after careful consideration, we found that doing so would be counter-productive. This is because the three instructions are quite distinct, and combining them into tabs would result in a lot of text being cramped together, leading to a loss of clarity. In my opinion, it would be more logical to present the instructions separately.

I will however update the _Linux_ part to be more specific.

clalancette reviewed (changes requested):
I've left some comments inline.

I also wanted to address this bit:

> If this is accepted, we plan to add more tutorial pages in the future.

That's totally fine, as long as the tutorials have to do with Webots/ROS integration.

ygoumaz committed:
Restore Simulation-Main

ygoumaz committed:
Update source/Tutorials/Advanced/Simulation-Webots/Installation-Linux.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Change linux to Ubuntu

ygoumaz committed:
Merge branch 'rolling' of https://github.com/cyberbotics/ros2_documentation into rolling

clalancette commented on source/Tutorials/Advanced.rst:
```diff
@@ -10,5 +10,5 @@ Advanced
    Advanced/Recording-A-Bag-From-Your-Own-Node-CPP
    Advanced/Recording-A-Bag-From-Your-Own-Node-Py
-   Advanced/Simulators/Simulation-Main
+   Advanced/Simulation-Main
```
```suggestion
   Advanced/Simulators/Simulation-Main
```

If we do that, then it is unnecessary to move at least some of the file below.

Another alternative might be to make additional subdirectories below `Simulators`; so instead of having `Simulators-Webots`, you would have `Simulators/Webots`.  I think that would better reflect what we are trying to do here.

ygoumaz replied:
I agree with you that it makes more sense with this file structure. However, it is inconsistent to use _Simulators_ as directory name when the corresponding HTML title is _Simulation_. I followed your suggestion but renamed the folder.

ygoumaz committed:
New folder level

ygoumaz committed:
minor

omichel commented:
@clalancette: I believe this PR is ready for your review again.

omichel commented:
@clalancette: could you please review this PR?

ygoumaz committed:
Merge branch 'ros2:rolling' into rolling

ygoumaz committed:
Migrate custom plugin tutorials to official documentation (#10)

* Improve python tutorial

* Add c++ to introduction step

* Add c++ for first part of implementation

* fixes

* fix

* tabs in note

* Update obstacle avoidance with c++

* fix whitespaces

* minor

* Update demos link

* Update source/Tutorials/Advanced/Simulation/Webots/Setting-Up-Simulation-Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulation/Webots/Setting-Up-Simulation-Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* Update source/Tutorials/Advanced/Simulation/Webots/Setting-Up-Simulation-Webots.rst

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

* General plugin

* add information on plugin def

* Update Setting-Up-Simulation-Webots.rst

---------

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

ygoumaz commented:
I have extended the PR with additional changes and improvements to the `Setup-Simulation-Webots` tutorial. New tabs allow to switch between Python and C++ implementation. There is also more detailed information about available devices and using custom plugins in Webots simulations.

omichel commented:
@clalancette and @kscottz: we would love to test this documentation during the Iron Irwini testing and tutorial party on May 3rd. So I hope it could be reviewed and merged by then.

clalancette reviewed (changes requested):
At the end of the day, what is keeping me from approving this is the rearrangement of the directory structure and things that are not related to Webots.

For instance, while there is a distinction between `Simulators` and `Simulation`, I don't think it is a big enough difference to change it and then have to have `redirect-from` links everywhere.  So I would still suggest not to change that.

Similarly, I would not add the new `Gazebo` page (I know that Kat had asked for this, but I agree with @omichel that this isn't the place to do this).

Essentially, just keep this PR focused on updating the Webots tutorials, and I think we can get it in.  The rest of it we'll have to do separately.

ygoumaz committed:
rename Simulation folder

ygoumaz committed:
Remove webots-gazebo comparison

ygoumaz commented:
There you go! I changed the folder name back to `Simulators` to avoid redirects. Instead, I renamed the title of the page, as it is less harmful to the structure and other files, but it makes much more sense to match it to the folder name.
I also removed the section comparing Webots to Gazebo.

omichel commented:
Thanks @clalancette for the review. @ygoumaz pushed the changes you requested. Could you please review it again?
Hopefully we can get this PR merged before the end of the Iron Testing Party so that we can address https://github.com/osrf/ros2_test_cases/issues/1025 on time.

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+Installation (macOS)
+======================================
```
```suggestion
Installation (macOS)
====================
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+----------
+
+The ``webots_ros2`` package provides an interface between ROS 2 and Webots.
+It includes several sub-packages, including ``webots_ros2_driver``, which allows to start Webots and communicate with it.
```
```suggestion
It includes several sub-packages, including ``webots_ros2_driver``, which allows you to start Webots and communicate with it.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+Tasks
+-----
+
+On macOS, a solution based on UTM virtual machines allows to improve the user experience with ROS 2 compared to native macOS installation, as it allows to run ROS in a Linux environment.
```
```suggestion
On macOS, a solution based on UTM virtual machines allows for an improved user experience with ROS 2 compared to native macOS installation, as it allows to run ROS in a Linux environment.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+On macOS, a solution based on UTM virtual machines allows to improve the user experience with ROS 2 compared to native macOS installation, as it allows to run ROS in a Linux environment.
+However, Webots should be installed natively on macOS and it will be able to communicate with the ROS nodes running in the Virtual Machine (VM).
+This solution allows for native 3D hardware acceleration for Webots.
+The VM is responsible to run all the ROS part (including RViz) and connects to the host machine through TCP connection to start Webots.
```
```suggestion
The VM is runs all the ROS part (including RViz) and connects to the host machine through TCP to start Webots.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+The VM is responsible to run all the ROS part (including RViz) and connects to the host machine through TCP connection to start Webots.
+A shared folder allows the script to transfer the world and other resource files from the VM to macOS where Webots is running.
+
+The following steps explain how to create the VM image with the installation of the ``webots_ros2`` released package. It is also possible to install it from sources.
```
```suggestion
The following steps explain how to create the VM image with the installation of the ``webots_ros2`` released package.
It is also possible to install it from sources.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+2 Configure the VM
+^^^^^^^^^^^^^^^^^^
+In this section, ROS 2 is installed in the VM and the shared folder is configured.
+The following instruction and command are all run inside the VM.
```
```suggestion
The following instructions and commands are all run inside the VM.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+The following instruction and command are all run inside the VM.
+
+* Open a terminal in the started VM and install the ROS 2 distribution you need by following the instructions in :doc:`../../../../Installation/Ubuntu-Install-Debians`:
+* Create a folder in the VM to use as shared folder.
```
```suggestion
* Create a folder in the VM to use as a shared folder.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst:
```diff
@@ -0,0 +1,231 @@
+3 Install ``webots_ros2``
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can then either install ``webots_ros2`` from the official released package, or install it from the last up-to-date sources committed on `Github <https://github.com/cyberbotics/webots_ros2>`_.
```
```suggestion
You can either install ``webots_ros2`` from the official released package, or install it from the latest up-to-date sources from `Github <https://github.com/cyberbotics/webots_ros2>`_.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-Ubuntu.rst:
```diff
@@ -0,0 +1,131 @@
+
+1 Install ``webots_ros2``
+^^^^^^^^^^^^^^^^^^^^^^^^^
+You can either install the official released package, or install it from the last up-to-date sources committed on `Github <https://github.com/cyberbotics/webots_ros2>`_.
```
```suggestion
You can either install the official released package, or install it from the latest up-to-date sources from `Github <https://github.com/cyberbotics/webots_ros2>`_.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-Windows.rst:
```diff
@@ -0,0 +1,197 @@
+1 Install WSL2
+^^^^^^^^^^^^^^^
+
+On Windows, WSL (Windows Subsystem for Linux) allows to improve the user experience with ROS 2 compared to native Windows installation, as it runs on a Linux platform.
```
```suggestion
On Windows, WSL (Windows Subsystem for Linux) improves the user experience with ROS 2 compared to native Windows installation, as it runs on a Linux platform.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-Windows.rst:
```diff
@@ -0,0 +1,197 @@
+
+3 Install ``webots_ros2``
+^^^^^^^^^^^^^^^^^^^^^^^^^
+You can then either install ``webots_ros2`` from the official released package, or install it from the last up-to-date sources committed on `Github <https://github.com/cyberbotics/webots_ros2>`_.
```
```suggestion
You can then either install ``webots_ros2`` from the official released package, or install it from the latest up-to-date sources from `Github <https://github.com/cyberbotics/webots_ros2>`_.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-Windows.rst:
```diff
@@ -0,0 +1,197 @@
+
+            source install/local_setup.bash
+
+
+
```
Nit: unnecessary spaces.

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Installation-Windows.rst:
```diff
@@ -0,0 +1,197 @@
+4 Launch the ``webots_ros2_universal_robot`` example
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+WSL doesn't support hardware acceleration (yet). Therefore, Webots should be started on Windows, while the ROS part is running inside WSL.
```
```suggestion
WSL doesn't support hardware acceleration (yet).
Therefore, Webots should be started on Windows, while the ROS part is running inside WSL.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+        .. code-block:: console
+
+            ros2 pkg create --build-type ament_python --node-name my_robot_driver my_package --dependencies rclpy geometry_msgs webots_ros2_driver
```
To avoid a warning from `ros2 pkg` nowadays, we should add in the `--license` flag:

```suggestion
            ros2 pkg create --build-type ament_python --license Apache-2.0 --node-name my_robot_driver my_package --dependencies rclpy geometry_msgswebots_ros2_driver
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+        .. code-block:: console
+
+            ros2 pkg create --build-type ament_cmake --node-name MyRobotDriver my_package --dependencies rclcpp geometry_msgs webots_ros2_driver pluginlib
```
```suggestion
            ros2 pkg create --build-type ament_cmake --license Apache-2.0 --node-name MyRobotDriver my_package --dependencies rclcpp geometry_msgs webots_ros2_driver pluginlib
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+5 Create the launch file
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let's create now the launch file to easily launch the simulation and the ROS controller with a single command.
```
```suggestion
Let's create the launch file to easily launch the simulation and the ROS controller with a single command.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+    .. group-tab:: Python
+
+        Finally, before you can start the launch file, you have to modify the ``setup.py`` file to include the extra files you added.
```
```suggestion
        Before you can start the launch file, you have to modify the ``setup.py`` file to include the extra files you added.
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+    .. group-tab:: C++
+
+        Finally, before you can start the launch file, you have to modify ``CMakeLists.txt`` and ``my_robot_driver.xml`` files:
```
```suggestion
        Before you can start the launch file, you have to modify ``CMakeLists.txt`` and ``my_robot_driver.xml`` files:
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+        .. literalinclude:: Code/my_robot_driver.xml
+            :language: xml
+
+        Finally, open ``my_package/CMakeLists.txt`` and replace its contents with:
```
```suggestion
        Open ``my_package/CMakeLists.txt`` and replace its contents with:
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+        .. literalinclude:: Code/CMakeLists.txt
+            :language: cmake
+
+        It exports the plugin configuration file with the ``pluginlib_export_plugin_description_file()`` command.
+
+        Next, it defines a shared library of your C++ plugin ``src/MyRobotDriver.cpp``.
+        It sets the ``include`` directories to include and specifies the library dependencies using ``ament_target_dependencies()``.
```
```suggestion
        The CMakeLists.txt exports the plugin configuration file with the ``pluginlib_export_plugin_description_file()``, defines a shared library of the C++ plugin ``src/MyRobotDriver.cpp``, and sets the include and library dependencies using ``ament_target_dependencies()``
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+            :lines: 24-38
+
+
+10 Updating additional files
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
10 Updating additional files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+.. tabs::
+
+    .. group-tab:: Python
+        Edit ``setup.py`` and replace ``'console_scripts'`` with:
```
```suggestion
    .. group-tab:: Python
    
        Edit ``setup.py`` and replace ``'console_scripts'`` with:
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+11 Test the obstacle avoidance code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As in task ``7``, launch the simulation from a terminal in your ROS 2 workspace:
```
```suggestion
Launch the simulation from a terminal in your ROS 2 workspace:
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+    In case you want to learn how to create your own robot model in Webots, you can check this `tutorial <https://cyberbotics.com/doc/guide/tutorial-6-4-wheels-robot>`_.
+
+3 Edit the my_robot_driver plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
3 Edit the my_robot_driver plugin
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

clalancette commented on source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst:
```diff
@@ -0,0 +1,730 @@
+
+    More details on ``webots_ros2_driver`` and ``WebotsLauncher`` arguments can be found `on the nodes reference page <https://github.com/cyberbotics/webots_ros2/wiki/References-Nodes>`_.
+
+6 Edit additional files
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
6 Edit additional files
^^^^^^^^^^^^^^^^^^^^^^^
```

clalancette reviewed (changes requested):
Thank for revamping this; this structure is now much easier to look at and reason about, and we should be able to move forward with this.

I've left quite a number of fixes inline, but they are all pretty easy to fix.

The bigger thing I want to discuss here is the length of the `Setting-Up-Simulation-Webots.rst` file.  That tutorial is much longer than most of the rest of our tutorials.  At the very least, I think the `Time` on it should be increased (to at least 30 or 40 minutes).  However, I think it would be best if that giant tutorial was split up into (at least) two tutorials.  The first one would go up to section 7, and get the basic package working.  The second tutorial would continue from there and change it so that the robot could avoid obstacles.  Then each tutorial is shorter and easier to digest for readers.

Does that make sense?

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots/Installation-MacOS.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Update Installation-MacOS.rst

ygoumaz committed:
Update Installation-MacOS.rst

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Update Setting-Up-Simulation-Webots.rst

ygoumaz committed:
Update source/Tutorials/Advanced/Simulators/Webots/Setting-Up-Simulation-Webots.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

ygoumaz committed:
Update Setting-Up-Simulation-Webots.rst

ygoumaz committed:
Update Installation-Windows.rst

omichel committed:
Removed trailing whitespace

ygoumaz committed:
split tutorial in 2 parts

ygoumaz committed:
update names

ygoumaz commented:
I have applied all your suggestions and split the tutorial in two parts. I agree that the original one was very long. There is now a "Basic" and an "Advanced" part. I have updated the `Background`, `Pre-requisites`, `Summary` and `Next steps` sections to match the new structure.

clalancette reviewed (approved):
This looks fantastic now.  Thank you for iterating and for your patience while I got back to this review.

The only remaining open question before I merge this is what branches you want this backported to.  It can go into any or all of `iron`, `humble`, or `foxy`.

ygoumaz commented:
You can backport these changes to all three distros! Thank you.
